[add/update] Add json -> xml, modify json printing.

kyoshinaga opened the pull request:
- Add: parseJSON, loadJSONFile, toXML to JSONUtil.
- Modify: toJSON return formatted string.

Commit:
[add/update] Add json -> xml, modify json printing.

- Add: parseJSON, loadJSONFile, toXML to JSONUtil.
- Modify: toJSON return formatted string.

hiroshinoji commented on src/main/scala/jigg/util/JSONUtil.scala:
```diff
@@ -35,13 +42,13 @@ object JSONUtil {
   }
 
-  def toJSON(x: Any): StringBuilder = x match {
+  def toJSON(x: Any): String = x match {
```
I think it's better to segment this method into different methods for each argument type, e.g., `readJSON(file: String)` (or `file: File`) and `toJSON(node: Node)` (or `node: Elem`?). Generally a method receiving `Any` type and changing the behavior internally is not a good practice.

kyoshinaga commented:
Thank you very much for your comment.
Ok, I'll separate them into `readJSON` and `toJSON`.

Commit:
[update] Segment toJSON method into different methods.

-Add: readJSON(file: String) into JSONUtil.
-Modify: toJSON(x: Any) -> toJSON(x: Node).
-Delete: test for unsupported inputs.

Commit:
[update] Delete wasted line from JSONUtil.toJSONFromNode

hiroshinoji commented on src/main/scala/jigg/util/JSONUtil.scala:
```diff
@@ -35,27 +42,24 @@ object JSONUtil {
-    case x: Elem      => toJSONFromNode(x)
-    case _            => new StringBuilder("")
-  }
+  def readJSON(file: String): String = toJSONFromNode(XML.loadFile(file))
```
Sorry to complain again. Maybe `readJSON` here does not tell the meaning of the method correctly. Perhaps this method can be removed (and prohibit users to recover json from XML file directly in one method), which seems more safe?

kyoshinaga replied:
The `readJSON` method is not needed in this situation.
I'll remove it from `JSONUtil`

hiroshinoji commented on src/main/scala/jigg/util/JSONUtil.scala:
```diff
@@ -116,6 +120,64 @@ object JSONUtil {
+  def parseJSON(json: String): JValue = parse(json)
+
+  def loadJSONFile(name: String): JValue = {
+    val sb = new StringBuilder
```
Full string of the input file can be obtained efficiently by:

``` scala
IOUtil.openIterator(name).mkString()
```

`openIterator(filename: String)` returns the iterator of the contents of the file (each element is one line).
`.mkString("\n")` recovers the original file.

It is known that Scala's original IO library has lower quality.

kyoshinaga replied:
Sorry for not knowing it.
I'll be careful about using the original IO library from now on.

hiroshinoji commented on src/main/scala/jigg/util/JSONUtil.scala:
```diff
@@ -116,6 +120,64 @@ object JSONUtil {
+    os.write(toJSON(x))
+  }
+
+  def parseJSON(json: String): JValue = parse(json)
```
This method may be redundant and unnecessary.

kyoshinaga replied:
Ok, I'll remove it.

hiroshinoji commented on src/main/scala/jigg/util/JSONUtil.scala:
```diff
@@ -116,6 +120,64 @@ object JSONUtil {
+
+  def toXML(json: JValue): Node = {
+    implicit val formats = DefaultFormats
+    val jsonList = json.extract[Map[String, Any]].toList
```
I suggest not to convert this `Map` into `List`. See below.

hiroshinoji commented on src/main/scala/jigg/util/JSONUtil.scala:
```diff
@@ -116,6 +120,64 @@ object JSONUtil {
+    var hasChildFlag = false
+    var hasTextFlag = false
+
+    for (i <- x){
```
This style seems very procedural. Following code might be clearer.

``` scala
    // Type of x is Map[String, Any]
    val tagString = x get ".tag" // Option[Any], but this should always exist.
    val textString = x get "text" // This might be None.
    val children: List[Node] =
      x.get(".child").map { _.asInstanceOf[List[Map[String, Any]]]
        .map(generateXML) } getOrElse List()

    val attrs: Map[String, String] =
      x.filter { case (k, v) => k != ".tag" && k != "text" && k != ".child" }
        .map { case (k, v) => (k, v.toString) }

    val node = textString match {
      case Some(text) => <xml>{text}</xml>
      case None => <xml/>
    }
    val tagChanged = node.copy(label = tagString.get.toString)
    val childAdded = XMLUtil.addChild(tagChanged, children) // do nothing when children is empty
    XMLUtil.addAttributes(childAdded, attrs) // the same
```

This code does not use mutable objects like `var` or `ArrayBuffer` to make the code intuitive, though I've not tested the behavior.

kyoshinaga replied:
Thank you very much for your comment and code.
I'm going to use your code as a reference.

hiroshinoji commented on src/main/scala/jigg/util/JSONUtil.scala:
```diff
@@ -116,6 +120,64 @@ object JSONUtil {
+          case false => <xml></xml>.copy(label = tagString)
+        }
+      }
+      case false => <xml/>.copy(label = tagString)
```
Unnecessary to catch cases where `hasTextFlag = true` but `hasChildFlag = false`? (seems theoretically possible)

kyoshinaga replied:
This part is my mistake.
But, it is removed when above code is implemented into `generateXML`.

hiroshinoji commented on src/main/scala/jigg/util/JSONUtil.scala:
```diff
@@ -116,6 +120,64 @@ object JSONUtil {
 
   def writeToJSON[T <: Node](x: T, os: BufferedWriter): Unit = {
-    os.write(toJSON(x).toString)
+    os.write(toJSON(x))
```
`toJSON` is a public method, so seems that this method is not generic enough to be implemented as a separate method.

kyoshinaga replied:
This separation seems to be not necessary.
I'll remove them from `JSONUtil` , and implement the output sequences into `Pipeline.run` and `Pipeline.runFromStdin`

Commit:
[update] Modify JSONUtil, and separate XMLParse method from JSONUtil.

-Modify: JSONUtil.toJSON, toXML.
-Add: hasChild, getChildNode, getAttributionList into XMLUtil.
-Add: Steps to output JSON file into Pipeline.run, runFromStdin.

hiroshinoji commented:
LGTM. Thank you!
